impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java

```diff
@@ -192,32 +191,26 @@ public TestcontainersKafkaCluster(DockerImageName kafkaImage, DockerImageName zo
         clusterConfig.getBrokerConfigs(() -> this).forEach(holder -> nodes.put(holder.getBrokerNum(), buildKafkaContainer(holder)));
     }
 
-    private static void validateClusterConfig(KafkaClusterConfig clusterConfig) {
-        if (Boolean.TRUE.equals(clusterConfig.getKraftMode()) && clusterConfig.getBrokersNum() < clusterConfig.getKraftControllers()) {
-            throw new IllegalStateException(
-                    "Due to https://github.com/ozangunalp/kafka-native/issues/88 we can't support controller only nodes so we need to fail fast. This cluster has "
-                            + clusterConfig.getBrokersNum() + " brokers and " + clusterConfig.getKraftControllers() + " controllers");
-        }
-    }
-
     @NotNull
     private KafkaContainer buildKafkaContainer(KafkaClusterConfig.ConfigHolder holder) {
         String netAlias = "broker-" + holder.getBrokerNum();
+        Properties properties = new Properties();
+        properties.putAll(holder.getProperties());
+        properties.put("log.dir", getBrokerLogDirectory(holder.getBrokerNum()));
```

Comment:
Why is this set here a not in the holder? Why is it special in test containers? Shouldn't each container have the same path?

Reply:
TestContainers has a mounted in volume `/kafka` so that the log dir can outlive the container getting restarted, so for broker-0 it's logdir is set to `/kafka/broker-0/`.

The InVm cluster creates a temp dir on the local machine and sets `log.dir` to `${tempDir}/broker-0`.

Reply:
Ok, but shouldn't the log dir still be consistent between them and thus set in the config holder?

Reply:
Having a go at refactoring now.

Reply:
Have pushed up a refactor that centralizes generation of the per-broker log dir configuration. 

I'm not sure if it's an improvement, there was an interface `KafkaEndpoints` being used to describe the public endpoints, which is an environment specific behaviour. So I tried to generalize this to describe other characteristics of the Cluster when run with a particular execution mode. I called it `MaterializedKafkaCluster`, but this doesn't feel great, so any advances on that would help. 

I also considered adding more interfaces like KafkaEndpoints (something like KafkaStorage) but then there's a lot of signatures that would need both things and it feels like they should be packaged up somehow.

As per the other chain, if this is too deep I'll revert and we can improve it separately.

Reply:
The refactoring definitely represents an improvement.